File: (2) - Add Two Numbers.py
# 72 ms
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        temp = ListNode(0)
        result = temp
        p = l1
        q = l2
        carry = 0
        while not (p is None and q is None):
            x = p.val if p is not None else 0
            y = q.val if q is not None else 0
            sum = x + y + carry
            carry = sum//10
            result.next = ListNode(sum % 10)
            result = result.next    
            p = p.next if p is not None else None
            q = q.next if q is not None else None
        if carry:
            result.next = ListNode(carry)
        return temp.next

def createListNodefromList(l: list):
    tempNode = ListNode(0)
    node = tempNode
    for element in l:
        node.next = ListNode(element)
        node = node.next
    return tempNode.next

def printListNode(listnode: ListNode):
    result = []
    while listnode:
        result.append(listnode.val)
        listnode = listnode.next
    print(result)

a = createListNodefromList([2,4,3])
b = createListNodefromList([5,6,4])
c = createListNodefromList([0])
d = createListNodefromList([0])
e = createListNodefromList([9,9,9,9,9,9,9])
f = createListNodefromList([9,9,9,9])

printListNode(Solution.addTwoNumbers(Solution, a, b))
printListNode(Solution.addTwoNumbers(Solution, c, d))
printListNode(Solution.addTwoNumbers(Solution, e, f))

File: 1 - 99/(2) - Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = (l1, l2) => {
    let head = new ListNode(), 
        result = head,
        first = l1,
        second = l2,
        carry = 0;
    
    while (first || second) {
        var sum = (first?.val ?? 0) + (second?.val ?? 0) + carry;

        result = result.next = new ListNode(sum % 10);
        carry = sum > 9 ? 1 : 0;

        first = first?.next;
        second = second?.next;
    }

    if (Boolean(carry)) result.next = new ListNode(1);
    
    return head.next;
};